### ISSUE TITLE
pnext: ScErrUnexpMetadata (SCERR7035): Internal error in unmanaged meta-data population

### ISSUE THREAD

**Issue by dan31** [Starcounter/Starcounter#3359]
After starting and stopping some apps, but not changing the class schemas, the database is in the following state at an attempt to start a database with no apps autostarted:
```
System.Exception: ScErrUnexpMetadata (SCERR7035): Internal error in unmanaged meta-data population.
Version: 2.3.797.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR7035.
   at Starcounter.SqlProcessor.SqlProcessor.MetalayerThrowIfError(UInt32 err) in C:\TeamCity\TeamCity9\buildAgent\work\sc-53\Level1\src\Starcounter\SqlProcessor\SqlProcessor.cs:line 282
   at Starcounter.SqlProcessor.SqlProcessor.PopulateRuntimeMetadata() in C:\TeamCity\TeamCity9\buildAgent\work\sc-53\Level1\src\Starcounter\SqlProcessor\SqlProcessor.cs:line 181
   at Starcounter.Hosting.StarcounterPackage.UpdateDatabaseSchemaAndRegisterTypes(String fullAppId, TypeDef[] unregisteredTypeDefs, TypeDef[] allTypeDefs) in C:\TeamCity\TeamCity9\buildAgent\work\sc-53\Level1\src\Hosting\Starcounter.Hosting\StarcounterPackage.cs:line 43
   at Starcounter.Hosting.Package.ProcessWithinCurrentApplication(Application application) in C:\TeamCity\TeamCity9\buildAgent\work\sc-53\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 261
   at Starcounter.Hosting.Package.Process() in C:\TeamCity\TeamCity9\buildAgent\work\sc-53\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 143
   at Starcounter.Hosting.Package.Process(IntPtr hPackage) in C:\TeamCity\TeamCity9\buildAgent\work\sc-53\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 56
   at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in C:\TeamCity\TeamCity9\buildAgent\work\sc-53\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 90
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR7035
```
Uploaded are three apps used and a database entire repository: http://rghost.net/8nsVw2qnC
CC @eriohl @k-rus @bigwad

**Comment by dan31** [comment]
I have the same situation on the other machine. Nothing is changing in a schema, only a starting and stopping applications. At some moment the "Bulker" application starts to spin infinitely, after which the issue takes place. To let the spin happen, I start "Bulker" by Ctrl+F5 in Release build from VS.

**Comment by alexanderstc** [comment]
I ran metalayer_functest on the uploaded database, it gives an error. Investigating.

**Comment by alexanderstc** [comment]
Looks like TODO on line 60 of level0/src/metalayer/create_metatable.cpp needs to be implemented somehow to allow TABLE_INFO::load_indexes() to recover existing layout id from indices,
if it's still possible after, i guess, #3278.
@eriohl could you comment, please.

**Comment by eriohl** [comment]
No it is no longer possible since there is no reference to a layout in the index. See comments in #3278.

**Comment by alexanderstc** [comment]
Thanks, I'll fix the metalayer then.

**Comment by k-rus** [comment]
@dan31 
It would be nice to add your case as a regression test. Can you provide reproduction steps and links to the source code of the apps? My idea is to have a test, which starts a database, apps in it and populate with necessary data if needed, then it stops the database and starts it again and then starts what is necessary to reproduce the error.

**Comment by alexanderstc** [comment]
@dan31 frankly, I don't see a possibility of a regression per se for this case.
The current state is the metalayer is out of sync with the current kernel design.
The metalayer is, simply, wrong. It didn't follow the kernel changes in #3278.
I'll fix it, the code will change notably, this case will never return.

**Comment by alexanderstc** [comment]
This isn't some subtle user data - driven case which has to be ambushed and trapped.

**Comment by k-rus** [comment]
> @dan31 frankly, I don't see a possibility of a regression per se for this case.
> The current state is the metalayer is out of sync with the current kernel design.
> The metalayer is, simply, wrong. It didn't follow the kernel changes in #3278.
> I'll fix it, the code will change notably, this case will never return.

This does not explain why the test cannot be done. 

The purpose of the test is to catch such mismatch that it is caught while running test build and before pushing to main branch. If there is a test, which allows to reproduce the error, then it is good to have it in the regression.

**Comment by dan31** [comment]
@k-rus @alexanderstc I should conclude now the case is likely reproducible. I will spend more time in running those examples with the current version and report to you if I find a way to get a simple regression test from it.

**Comment by alexanderstc** [comment]
@k-rus @dan31
We have this intention: https://git.starcounter.org/BlueStar/level0/issues/50
"Multi-threaded and memory leak test of SQL parser".
Perhaps we can also somehow make a simulated multi-app test for metalayer and sql parser,
the test should be runnable on level0 ideally, as easy as existing metalayer_functest.

**Comment by alexanderstc** [comment]
I looked a bit at the level1 code and it seems to me that star_prepare_system_tables() can be called
from multiple threads simultaneously, though the function is definitely not thread-safe.
I mutexed the function and pushed it to pnext.
@dan31 if you have a minute, could you test from latest pnext and see if the bug reproduces?

**Comment by k-rus** [comment]
> I looked a bit at the level1 code and it seems to me that star_prepare_system_tables() can be called
> from multiple threads simultaneously, though the function is definitely not thread-safe.
> I mutexed the function and pushed it to pnext.

This function can be called only from one thread. Bootstrapper has logic which guarantees single working thread at the moment. Thus no need to extra code for thread safety. Furthermore, in future it will be called during database creation.

**Comment by k-rus** [comment]
> "Multi-threaded and memory leak test of SQL parser".
> Perhaps we can also somehow make a simulated multi-app test for metalayer and sql parser,
> the test should be runnable on level0 ideally, as easy as existing metalayer_functest.

In my mind it is more like unit test, since SQL parser does not require a database to be created. I have such test in develop, which is implemented in Level1.
Is it  still possible to call the parser without need in a database?

**Comment by alexanderstc** [comment]
@k-rus
> Is it still possible to call the parser without need in a database?
yes, we have sqlparser_unittest, which works like that.

**Comment by k-rus** [comment]
> Is it still possible to call the parser without need in a database?
> yes, we have sqlparser_unittest, which works like that.

Then the test for this issue is not related to https://git.starcounter.org/BlueStar/level0/issues/50

**Comment by alexanderstc** [comment]
What meant was we could have a similar multi-tread & leak test for the metalayer.

**Comment by k-rus** [comment]
> What meant was we could have a similar multi-tread & leak test for the metalayer.

It is off-topic to this issue. Having such test is good idea, but should be discussed in a separate issue.

**Comment by dan31** [comment]
@alexanderstc, I tried it with ```v2.3.826```, I confirm it doesn't improve the situation. The way to reproduce is to try starting the code below from VS (F5 + Debug x64) atop empty database. First time it usually starts okay, next time it takes infinite time to start (race condition?), so that I have to kill processes. After that situation, I get same exception. That is an app I have:
```
using System;
using Starcounter;
using System.Threading;

namespace Bulker {
    [Database]
    public class Record {
        public ulong oid;
        public ulong oidSum;
        public string origin;
    }
    class Program {
        static ulong CalcSum(string origin) {
            var rs = Db.SQL("select r from Record r where r.origin=?", origin);
            ulong sum = 0;
            foreach (var r in rs)
                sum += r.GetObjectNo();
            return sum;
        }

        static ulong CalcCount(string origin) {
            return (UInt64) Db.SlowSQL<Int64>("select count(r) from Record r where r.origin=?", origin).First;
        }

        static bool Check(string origin) {
            ulong oidSum = 0;
            ulong maxOid = 0;
            var records = Db.SQL("select p from Record p where p.origin=?", origin);
            if (CalcCount(origin) == 0) return true;
            foreach (Record record in records) {
                if (record.oid > maxOid)
                    maxOid = record.oid;
                if (record.oid != record.GetObjectNo()) {
                    Console.WriteLine("No integrity on: localOID={0}, originOID={1}", record.GetObjectNo(), record.oid);
                    return false;
                }
                oidSum += record.GetObjectNo();
            }
            Record mr = (Record)DbHelper.FromID(maxOid);
            if (mr.oidSum != oidSum) {
                Console.WriteLine("No integrity on: localSum={0}, originSum={1}", oidSum, mr.oidSum);
            }
            Console.WriteLine("Integrity check for sum={0} maxOid={1} sucessful", oidSum, maxOid);
            return true;
        }

        static void Main() {
            Thread.Sleep(10000);
            string orig = "ground";
            ulong oidSum = CalcSum(orig);
            ulong cntr = CalcCount(orig);
            while (true) {
                for (int i = 1; i <= 10; ++i) {
                    Db.Transact(() => {
                        var p = new Record();
                        p.oid = p.GetObjectNo();
                        p.origin = orig;
                        oidSum += p.oid;
                        p.oidSum = oidSum;
                        cntr++;
                    });
                }
                Console.WriteLine("Inserted are {0} objects, sum = {1}", cntr, oidSum);
                Thread.Sleep(500);
                Check("vessel");
            }
        }
    }
}
```

@alexanderstc, whether you have hard time to reproduce the situation, ask me to give you access to a machine where I have it reproducible. @k-rus @alexanderstc I raise it to a showstopper bug, if you have nothing against, since I don't see how to get data from the database in the state.

**Comment by alexanderstc** [comment]
@dan31 I ran your sample on 2 machines - doesn't reproduce.
Please give me access to the reprducing machine.

**Comment by dan31** [comment]
@alexanderstc After the first time it occurred with a new version, I gave the new version a use for several hours more today, and could not get the same condition any more. I might have started and restarted the apps around 50 times during that use. Probably something was "cached" at the first run when condition occurred, yet that's not a satisfactory explanation (what might have gotten cached?), unfortunately I don't have any better explanations at hand today :8ball: . I can suggest the original "Bulker" code with ```Thread.Sleep``` usage in ```Main``` might have increased the likelihood of having the condition observed. However, currently I changed the code so that it uses ```Timer``` events and no ```Thread.Sleep``` on the ```Main```. With such a code, the condition isn't observed. I intentionally went back to the code with ```Thread.Sleep``` for a while, but in a run of 2 hours the condition haven't been observed either. Hence I remove the "Showstopper" here and keep an eye on the condition further on. @alexanderstc @k-rus thank you both for a detailed follow up.

**Comment by k-rus** [comment]
@dan31 
Your code cannot affect meta-data population, since it is executed afterwords. The only directions I can think of are restarting broken database, or refactoring of application schema or both.

**Comment by alexanderstc** [comment]
@dan31 also, as @k-rus pointed to me several times, sometimes level0 and level1 don't build correctly
after certain code change or repo pull. Sometimes deep cleaning of all built and generated files is needed.
Please also notice that the error in the database sample you uploaded is quite peculiar.
As if the second meta-table, the Token, is created with parent zero instead of MotherOfAllLayouts.
I can't easily explain how this can be achieved through the current metalayer code.

**Comment by alexanderstc** [comment]
Closing this as "won't reproduce".

**Comment by alexanderstc** [comment]
I'm still updating the metalayer to follow kernel #3278.

**Comment by k-rus** [comment]
See #3368

**Comment by k-rus** [comment]
> See #3368

No it is different code.

**Comment by dan31** [comment]
After running for a day, the problem reoccured:
```
System.Exception: ScErrUnexpMetadata (SCERR7035): Internal error in unmanaged meta-data population.
Version: 2.3.826.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR7035.
   at Starcounter.SqlProcessor.SqlProcessor.MetalayerThrowIfError(UInt32 err) in C:\TeamCity\TeamCity9\buildAgent\work\sc-56\Level1\src\Starcounter\SqlProcessor\SqlProcessor.cs:line 282
   at Starcounter.SqlProcessor.SqlProcessor.PopulateRuntimeMetadata() in C:\TeamCity\TeamCity9\buildAgent\work\sc-56\Level1\src\Starcounter\SqlProcessor\SqlProcessor.cs:line 181
   at Starcounter.Hosting.StarcounterPackage.UpdateDatabaseSchemaAndRegisterTypes(String fullAppId, TypeDef[] unregisteredTypeDefs, TypeDef[] allTypeDefs) in C:\TeamCity\TeamCity9\buildAgent\work\sc-56\Level1\src\Hosting\Starcounter.Hosting\StarcounterPackage.cs:line 43
   at Starcounter.Hosting.Package.ProcessWithinCurrentApplication(Application application) in C:\TeamCity\TeamCity9\buildAgent\work\sc-56\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 261
   at Starcounter.Hosting.Package.Process() in C:\TeamCity\TeamCity9\buildAgent\work\sc-56\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 143
   at Starcounter.Hosting.Package.Process(IntPtr hPackage) in C:\TeamCity\TeamCity9\buildAgent\work\sc-56\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 56
   at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in C:\TeamCity\TeamCity9\buildAgent\work\sc-56\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 86
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR7035
```
CC @linkdata

**Comment by linkdata** [comment]
Do you need to look at the machine state? We get this error when trying to start the database, so it may be reproducible by copying the database files.

**Comment by linkdata** [comment]
I've made copies of the databases involved. Unless you need access to the live machine, I'm going to go ahead and nuke the installation so I can proceed with other testing on it.

**Comment by alexanderstc** [comment]
@dan31 I would appreciate access to a machine where this is reproducible.

**Comment by linkdata** [comment]
Just verified that copying the database to another installation and trying to start it there gives the same exception, so I will go ahead and wipe the original.

@alexanderstc, I will send you the details on where to find the database.

**Comment by alexanderstc** [comment]
@linkdata I already have a copy of such a database, which @dan31 uploaded earler.
The error itself is consequential, I'd like to debug the steps which lead to the error.

**Comment by linkdata** [comment]
Ok, holding off.

**Comment by alexanderstc** [comment]
@bigwad , as you suggested, I renamed the optlog on this error db @dan31 uploaded earlier, and the error is gone.
> Uploaded are three apps used and a database entire repository: http://rghost.net/8nsVw2qnC

So, run metalaer_functest on this db with optlog - gives error because Token meta-table's parent is zero instead of MotherOfAllLayouts,

Run on the same db with optlog removed - no error.

Please have a look.

CC @eriohl

**Comment by dan31** [comment]
I got the situation again, and @bigwad way worked well to reproduce the issue. When you remove logopt file, database starts well. When you bring back the logopt file, the database won't start. I don't run any test, just try to start a database from Administrator. Thanks @bigwad @alexanderstc for reproduction. I guess the question is still how to get database to such a state. However, we now have binaries which allow to reproduce the situation.

**Comment by alexanderstc** [comment]
This is from @linkdata over e-mail:
> I think the corruption may be related to Bulker app pushing invoking Scheduling.ScheduleTask() faster than SC can process them.
> I'm not sure if our behavior is well-defined when this happens.
> If that is the cause, it should be easy to provoke it, and should be reproducible on any machine.

**Comment by linkdata** [comment]
Well, that was before we identified log optimization as the likely culprit. I no longer think Scheduling.ScheduleTask() overload is the cause of the error.

**Comment by eriohl** [comment]
I'll check it. Could you upload the database to somewhere that isn't blocked by the browser however (like our dropbox account.)

**Comment by alexanderstc** [comment]
> Uploaded are three apps used and a database entire repository: http://rghost.net/8nsVw2qnC

Can some1 upload this to buildserver ftp or something? I don't have rights to do it.

**Comment by k-rus** [comment]
> Uploaded are three apps used and a database entire repository: http://rghost.net/8nsVw2qnC

I uploaded the file to @eriohl's exchange folder on dropbox

**Comment by eriohl** [comment]
Great. Thank you

**Comment by eriohl** [comment]
Yes you where right. logopt drops the reference to the base layout. Fixed this but I discovered some additional problems with the optlog load that I'm looking into as well.

**Comment by eriohl** [comment]
There was more problems with base layouts not being handled properly in logopt. Problem should now be resolved now. All previously created optlog files are invalid however.
